File: src/main/java/rag/trace/JsonlTraceSink.java
package main.java.rag.trace;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class JsonlTraceSink implements TraceSink {
    private final Path logFile;
    
    public JsonlTraceSink(Path logsDir) {
        LocalDateTime now = LocalDateTime.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss");
        String fileName = "run-" + now.format(formatter) + ".jsonl";
        
        this.logFile = logsDir.resolve(fileName);
        
        try {
            Files.createDirectories(logsDir);
        } catch (IOException e) {
            throw new RuntimeException("Failed to create logs directory: " + logsDir, e);
        }
    }
    
    @Override
    public void record(TraceEvent event) {
        try {
            String jsonLine = event.toJson() + "\n";
            Files.write(logFile, jsonLine.getBytes(), 
                       StandardOpenOption.CREATE, 
                       StandardOpenOption.APPEND);
        } catch (IOException e) {
            throw new RuntimeException("Failed to write trace event to: " + logFile, e);
        }
    }
    
    public Path getLogFile() {
        return logFile;
    }
}

File: iteration1_py/src/trace/jsonl_trace_sink.py
import json
from pathlib import Path
from datetime import datetime

from src.trace.trace_sink_abstract import TraceSinkAbstract
from src.trace.trace_event import TraceEvent


class JsonlTraceSink(TraceSinkAbstract):
    """
    Trace sink that writes events to JSONL (JSON Lines) format.
    One JSON object per line.
    """
    
    def __init__(self, logs_dir: Path):
        """
        Initialize JsonlTraceSink.
        
        Args:
            logs_dir: Directory to write log files
        """
        logs_dir = Path(logs_dir)
        
        # Create logs directory if it doesn't exist
        try:
            logs_dir.mkdir(parents=True, exist_ok=True)
        except Exception as e:
            raise RuntimeError(f"Failed to create logs directory: {logs_dir}") from e
        
        # Generate log file name with timestamp
        now = datetime.now()
        timestamp = now.strftime("%Y%m%d-%H%M%S")
        self.log_file = logs_dir / f"run-{timestamp}.jsonl"
    
    def record(self, event: TraceEvent) -> None:
        """
        Record a trace event to JSONL file.
        
        Args:
            event: TraceEvent to record
        """
        try:
            json_line = event.to_json() + "\n"
            with open(self.log_file, 'a', encoding='utf-8') as f:
                f.write(json_line)
        except Exception as e:
            raise RuntimeError(f"Failed to write trace event to {self.log_file}") from e
    
    def get_log_file(self) -> Path:
        """Get the log file path."""
        return self.log_file
